source/row_selector.ts

```diff
@@ -0,0 +1,316 @@
+import { ArrayTableModel } from './array_table_model';
+import { TableModel } from './table_model';
+import { AddRowOperation, MoveRowOperation, Operation }
+  from './operations';
+
+/** Provides the functionality needed to resize a table's columns. */
+export class RowSelector {
+
+  /** Constructs a ColumnResizer.
+   * @param table - The interface to the table being resized.
+   */
+  constructor(table: TableModel) {
+    this.isShiftDown = false;
+    this.isCtrlDown  = false;
+    this.isMouseDown = false;
+    this.isUpDown = false;
+    this.isDownDown  = false;
+    this.isAdding = true;
+    this.selectedRows = new ArrayTableModel();
+    for(let i = 0; i < table.rowCount; ++i) {
+      if(i === 0) {
+        this.selectedRows.addRow([true]);
+      } else {
+        this.selectedRows.addRow([false]);
+      }
+    }
+    this.currentRow = 0;
+    this.hilightedRow = 0;
+    this.previousRow = 0;
+    table.connect(this.handleOperations.bind(this));
+    this.s0();
+  }
+
+  public isSelected(row: number): boolean {
+    if(this.selectedRows.get(row, 0) === true) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /** Handles the cursor moving. */
+  public onMouseEnter(row: number) {
+    if(this.isMouseDown) {
+      this.currentRow = row;
+      if(this.state === 2) {
+        this.s2();
+      } else if(this.state === 4) {
+        this.s4();
+      } else if(this.state === 6) {
+        this.s6();
+      }
+    }
+  }
+
+  /** Handles pressing down a mouse button. */
+  public onMouseDown(event: PointerEvent, row: number) {
+    if(event.button === 0) {
+      this.isMouseDown = true;
+      this.currentRow = row;
+      if(this.state === 0) {
+        this.s0();
+      } else if(this.state === 2) {
+        this.s2();
+      }
+    }
+  }
+
+  /** Handles releasing a mouse button. */
+  public onMouseUp(event: PointerEvent) {
+    if(event.button === 0) {
+      this.isMouseDown = false;
+      if(this.state === 2) {
+        this.s2();
+      } else if(this.state === 4) {
+        this.s0();
+      } else if(this.state === 6) {
+        this.s0();
+      }
+    }
+  }
+
+  /** Handles a keyboard button being pressed down. */
+  public onKeyDown(event: KeyboardEvent) {
+    const keyCode = event.keyCode;
+    if(keyCode === 38) {
+      this.isUpDown = true;
+      if(this.currentRow > 0) {
+        this.currentRow--;
+      }
+      if(this.state === 0) {
+        this.s0();
+      } else if(this.state === 2) {
+        this.s2();
+      } else if(this.state === 4) {
+        this.s4();
+      } else if(this.state === 7) {
+        this.s7();
+      }
+    } else if(keyCode === 40) {
+      this.isDownDown = true;
+      if(this.currentRow < this.selectedRows.rowCount - 1) {
+        this.currentRow++;
```

Comment:
```++x``` instead of ```x++```, feel free to read about the subtle but important difference between preincrement and postincrement.